cpp/parquetjni/parquetjni.cc

```diff
@@ -229,46 +262,85 @@ class DatasetBatchReader : public arrow::RecordBatchReader {
     // Access private fields
     DatasetBatchReader *reader = static_cast<DatasetBatchReader *>(out->get());
     // Load the first file and get the schema.
-    ARROW_RETURN_NOT_OK(reader->LoadNext());
-    reader->schema_ = reader->current_reader_->schema();
+    ARROW_RETURN_NOT_OK(reader->EnqueueReads(1));
+    if (reader->file_readers_.empty()) {
+      return arrow::Status::Invalid("No files in dataset");
+    }
+    TRACE("ParquetJniDataset::Open::LoadNext", profile_start, load_end);
+    std::shared_ptr<arrow::RecordBatchReader> batch_reader;
+    RETURN_NOT_OK(reader->file_readers_.front().get().GetReader(&batch_reader));
+    reader->schema_ = batch_reader->schema();
+    TRACE("ParquetJniDataset::Open::Schema", load_end, open_end);
     return arrow::Status::OK();
   }
 
   arrow::Status LoadNext() {
-    current_reader_ = nullptr;
-    if (current_jni_reader_) {
-      RETURN_NOT_OK(current_jni_reader_->Close());
+    if (!file_readers_.empty()) {
+      VLOG(2) << "Closing a file";
+      BufferedFile &buffered_file =
+          const_cast<BufferedFile &>(file_readers_.front().get());
+      RETURN_NOT_OK(buffered_file.Close());
+      file_readers_.pop();
     }
-    current_jni_reader_ = nullptr;
-    if (file_queue_.empty()) {
-      return arrow::Status::OK();
+    return EnqueueReads(kMaxFilesBuffered);
+  }
+
+  arrow::Status EnqueueReads(size_t max_files) {
+    const auto profile_start = std::chrono::steady_clock::now();
+    while (file_readers_.size() < max_files && !file_queue_.empty()) {
+      const std::string next_file = file_queue_.front();
+      file_queue_.pop();
+      std::packaged_task<BufferedFile()> task(
+          [next_file, this, profile_start]() {
+            VLOG(2) << "Starting read of file: " << next_file;
+            ParquetJniReader *jni_reader;
+            arrow::Status status = SingleFileParquetJniReader::Open(
+                dataset_->endpoint_, dataset_->bucket_, next_file,
+                dataset_->auth_token_, &jni_reader);
+            if (!status.ok()) {
+              return BufferedFile{status, nullptr, nullptr};
+            }
+            std::shared_ptr<arrow::RecordBatchReader> batch_reader;
+            status = jni_reader->GetRecordBatchReader(
+                columns_, time_column_, min_utc_ns_, max_utc_ns_,
+                filter_row_groups_, &batch_reader);
+            if (!status.ok()) {
+              return BufferedFile{status, nullptr, nullptr};
+            }
+            TRACE_WITH("ParquetJniDataset::EnqueueReads", next_file, profile_start,
+                       load_end);
+            return BufferedFile(arrow::Status::OK(),
+                                std::unique_ptr<ParquetJniReader>(jni_reader),
+                                std::move(batch_reader));
+          });
+      std::future<BufferedFile> future = task.get_future();
+      std::thread thread(std::move(task));
```

Comment:
can we use threadpool so that we can reuse threads?

Reply:
C++ doesn't have a built in pool. I may convert this to use a Boost pool.

Reply:
I see. That's fine then.